### ISSUE TITLE
[strict-boolean-expressions] missing error when union includes array or object

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#4876]
### Bug Report

- __TSLint version__: 5.20.0
- __TypeScript version__: 3.6.3
- __Running TSLint via__: CLI

#### Reproduction using [TSLint Playground](https://palantir.github.io/tslint-playground)

For some reason this rule doesn't seem to error in the playground?

#### TypeScript code being linted

```ts
{
  declare const expression: boolean | string;
  if (expression) {} // Expect: error. Actual: error. ✅
}
{
  declare const expression: boolean | string[];
  if (expression) {} // Expect: error. Actual: no error! ❌
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-boolean-expressions": [true]
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
@OliverJAsh thanks for the bug report, but I can't reproduce this locally either. Switching the example to valid syntax for an `index.ts`:

```
  const expression1: boolean | string = Math.random() ? true : 'no';
  if (expression1) {
  } // Expect: error. Actual: error. ✅

  const expression2: boolean | string[] = Math.random() ? true : ['no', 'you'];
  if (expression2) {
  } // Expect: error. Actual: no error! ❌
```

Can you provide a working reproduction?

**Comment by silviogutierrez** [comment]
I was just having this issue today. Arrays are always truthy and yet, no complaints:

```
# package.json

{
  "dependencies": {
    "tslint": "^5.20.0",
    "typescript": "^3.6.4"
  }
}

# tsconfig.json
{
    "compilerOptions": {
        "strict": true
    },
    "include": ["./**/*"],
    "exclude": ["node_modules"]
}

# tslint.json
{
    "extends": ["tslint:recommended"],
    "jsRules": {},
    "rules": {
        "strict-boolean-expressions": true
    },
    "rulesDirectory": []
}

# test.ts

function this_is_correct_flagged(foo: string|boolean) {
    if (!foo) {
        return true;
    }
    return false;
}

function this_is_not(foo: string[]|boolean) {
    if (!foo) {
        return true;
    }
    return false;
}
```

Command to run: `node_modules/.bin/tslint -p . test.ts`.

**Comment by OliverJAsh** [comment]
@JoshuaKGoldberg In your example, the array has type `true | string[]` (despite the annotation). Try this:

```ts
const fn = (boolean: boolean) => {
    const expression2: boolean | string[] = Math.random() ? boolean : ['no', 'you'];
    if (expression2) {
    } // Expect: error. Actual: no error! ❌
}
```

(I originally used `declare const` to demonstrate this, which I think is valid syntax?)

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
